File: src/http.go
package main

import (
	"bytes"
	"crypto/tls"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"net/textproto"
	"net/url"
	"strings"
	"time"
)

// Cloudflare's edge blocks unrecognised User-Agents outright, with a 403 and
// "error code: 1010", so every request has to identify itself.
const userAgent = "komodoc/1.0"

// do makes one HTTP round trip, returning (status, body) rather than an error
// on a 4xx: both APIs here put their error detail in the response body.
func do(method, target string, headers map[string]string, body []byte, timeout time.Duration) (int, []byte) {
	var reader io.Reader
	if body != nil {
		reader = bytes.NewReader(body)
	}
	req, err := http.NewRequest(method, target, reader)
	if err != nil {
		die("%s %s: %v", method, target, err)
	}
	req.Header.Set("user-agent", userAgent)
	for name, value := range headers {
		req.Header.Set(name, value)
	}

	client := &http.Client{Timeout: timeout}
	response, err := client.Do(req)
	if err != nil {
		// A workers.dev hostname resolves as soon as it is created, but its TLS
		// certificate is issued a few minutes later. Until then the handshake is
		// refused, which is alarming and not actually a problem.
		var certErr *tls.CertificateVerificationError
		parsed, _ := url.Parse(target)
		if parsed != nil && strings.Contains(target, "workers.dev") &&
			(errors.As(err, &certErr) || strings.Contains(err.Error(), "tls:")) {
			die("TLS handshake refused by %s.\n\n"+
				"  A newly created workers.dev subdomain resolves before its\n"+
				"  certificate is issued. This usually clears within a few minutes.\n\n"+
				"  Wait for it with:\n"+
				"    until curl -sfo /dev/null %s://%s/; do sleep 30; done\n\n"+
				"  then try again.", parsed.Hostname(), parsed.Scheme, parsed.Hostname())
		}
		die("%s %s: %v", method, target, err)
	}
	defer response.Body.Close()

	raw, err := io.ReadAll(response.Body)
	if err != nil {
		die("%s %s: %v", method, target, err)
	}
	return response.StatusCode, raw
}

// postAuthed carries the GitHub token the CLI signed in with.
func postAuthed(target string, payload any, token string, timeout time.Duration) (int, map[string]any) {
	return postDecoded(target, payload, token, timeout)
}

// postDecoded is the body of both: encode, post, decode. An empty token sends
// no authorization header at all, which is what an unauthenticated call means
// -- a deployment whose publishers policy is "anyone" takes uploads with no
// bearer at all. Without a bearer, the server treats a request as
// cookie-authenticated and applies the cross-site checks in rule A, so the CLI
// carries the same marker header the browser shell does; a bearer-carrying
// call skips those checks regardless.
func postDecoded(target string, payload any, token string, timeout time.Duration) (int, map[string]any) {
	body, err := json.Marshal(payload)
	if err != nil {
		die("could not encode the request: %v", err)
	}
	headers := map[string]string{"content-type": "application/json", "x-komodoc-client": "cli"}
	if token != "" {
		headers["authorization"] = "Bearer " + token
	}
	status, raw := do("POST", target, headers, body, timeout)

	var decoded map[string]any
	if err := json.Unmarshal(raw, &decoded); err != nil {
		return status, map[string]any{"error": truncate(string(raw), 300)}
	}
	return status, decoded
}

type formPart struct {
	name        string
	filename    string
	contentType string
	body        string
}

// buildMultipart returns the Content-Type header and body for a
// multipart/form-data upload.
func buildMultipart(parts []formPart) (string, []byte) {
	var buffer bytes.Buffer
	writer := multipart.NewWriter(&buffer)
	for _, part := range parts {
		disposition := fmt.Sprintf(`form-data; name="%s"`, part.name)
		if part.filename != "" {
			disposition += fmt.Sprintf(`; filename="%s"`, part.filename)
		}
		header := textproto.MIMEHeader{}
		header.Set("Content-Disposition", disposition)
		header.Set("Content-Type", part.contentType)

		field, err := writer.CreatePart(header)
		if err != nil {
			die("could not build the upload: %v", err)
		}
		if _, err := io.WriteString(field, part.body); err != nil {
			die("could not build the upload: %v", err)
		}
	}
	if err := writer.Close(); err != nil {
		die("could not build the upload: %v", err)
	}
	return writer.FormDataContentType(), buffer.Bytes()
}

func truncate(text string, limit int) string {
	if len(text) <= limit {
		return text
	}
	return text[:limit]
}

File: komodoc/src/http.rs
//! One HTTP client for everything this binary asks of other servers: GitHub,
//! S3, and the deployment the command line publishes to.

use std::sync::OnceLock;
use std::time::Duration;

use serde_json::Value;

/// Every request identifies itself; some edges refuse an anonymous client.
pub const USER_AGENT: &str = "komodoc/1.0";

/// The shared client. Connection pooling is why it is one rather than one per
/// call; the timeout is per request, set where the request is made.
pub fn client() -> &'static reqwest::Client {
    static CLIENT: OnceLock<reqwest::Client> = OnceLock::new();
    CLIENT.get_or_init(|| {
        reqwest::Client::builder()
            .user_agent(USER_AGENT)
            .timeout(Duration::from_secs(300))
            .build()
            .expect("a client with no special needs builds")
    })
}

/// One round trip, returning (status, body) rather than an error on a 4xx:
/// both APIs this talks to put their error detail in the response body.
pub async fn send(
    method: reqwest::Method,
    target: &str,
    headers: &[(&str, &str)],
    body: Option<Vec<u8>>,
    timeout: Duration,
) -> Result<(u16, Vec<u8>), String> {
    let mut request = client().request(method.clone(), target).timeout(timeout);
    for (name, value) in headers {
        request = request.header(*name, *value);
    }
    if let Some(body) = body {
        request = request.body(body);
    }
    let response = request
        .send()
        .await
        .map_err(|err| format!("{method} {target}: {err}"))?;
    let status = response.status().as_u16();
    let raw = response
        .bytes()
        .await
        .map_err(|err| format!("{method} {target}: {err}"))?;
    Ok((status, raw.to_vec()))
}

/// Encode, post, decode. An empty token sends no authorization header at all,
/// which is what an unauthenticated call means -- a deployment whose
/// publishers policy is "anyone" takes uploads with no bearer at all. Without
/// a bearer, the server treats a request as cookie-authenticated and applies
/// the cross-site checks in rule A, so the CLI carries the same marker header
/// the browser shell does; a bearer-carrying call skips those checks
/// regardless.
pub async fn post_json(
    target: &str,
    payload: &Value,
    token: &str,
    timeout: Duration,
) -> Result<(u16, Value), String> {
    let body = serde_json::to_vec(payload)
        .map_err(|err| format!("could not encode the request: {err}"))?;
    let bearer = format!("Bearer {token}");
    let mut headers = vec![
        ("content-type", "application/json"),
        ("x-komodoc-client", "cli"),
    ];
    if !token.is_empty() {
        headers.push(("authorization", bearer.as_str()));
    }
    let (status, raw) = send(reqwest::Method::POST, target, &headers, Some(body), timeout).await?;
    Ok((status, decode(&raw)))
}

/// A plain GET, decoded when it is JSON.
pub async fn get_json(target: &str, timeout: Duration) -> Result<(u16, Value), String> {
    let (status, raw) = send(reqwest::Method::GET, target, &[], None, timeout).await?;
    Ok((status, decode(&raw)))
}

fn decode(raw: &[u8]) -> Value {
    serde_json::from_slice(raw).unwrap_or_else(
        |_| serde_json::json!({"error": truncate(&String::from_utf8_lossy(raw), 300)}),
    )
}

/// The error message out of an API reply, falling back to the whole reply when
/// it has no error field.
pub fn detail_of(payload: &Value) -> String {
    match payload.get("error") {
        Some(Value::String(message)) => message.clone(),
        Some(other) => other.to_string(),
        None => payload.to_string(),
    }
}

/// A string field of a decoded JSON object, or "" when it is not there.
pub fn text(value: &Value, field: &str) -> String {
    value
        .get(field)
        .and_then(Value::as_str)
        .unwrap_or_default()
        .to_string()
}

pub fn truncate(text: &str, limit: usize) -> String {
    if text.len() <= limit {
        return text.to_string();
    }
    let mut end = limit;
    while !text.is_char_boundary(end) {
        end -= 1;
    }
    text[..end].to_string()
}
